File: bindings/javascript/packages/core/src/fusion.ts
/**
 * Single-qubit gate-fusion DAG binding (C-side since v0.2.0, JS
 * binding since v0.4.7).
 *
 * Wraps ``src/optimization/fusion/fusion.{c,h}`` so JavaScript code
 * can build a symbolic circuit, run the run-length single-qubit
 * fuser over it, and execute the fused schedule on a
 * {@link QuantumState}.  Mirrors the Python ``moonlab.fusion``
 * (v0.4.4) and Rust ``moonlab::fusion`` (v0.4.6) surfaces.
 *
 * The fuser collapses adjacent single-qubit gates on the same qubit
 * into one 2x2 matrix product, dropping repeated full-state passes.
 * On a five-layer hardware-efficient ansatz at n = 16 the fused
 * execution is roughly 2.2x faster than the unfused dispatch; the
 * full performance discussion lives in
 * ``src/optimization/fusion/fusion.h``.
 *
 * @example
 * ```typescript
 * import { QuantumState, FusedCircuit } from '@moonlab/quantum-core';
 *
 * const circuit = await FusedCircuit.create(4);
 * circuit.h(0).rz(0, 0.3).rx(0, 0.7).cnot(0, 1).rz(1, 0.4);
 * const { fused, stats } = circuit.compile();
 * console.log(stats);
 *
 * const state = await QuantumState.create({ numQubits: 4 });
 * fused.execute(state);
 *
 * fused.dispose();
 * circuit.dispose();
 * state.dispose();
 * ```
 *
 * References:
 *  - T. Haener and D. S. Steiger, "0.5 Petabyte Simulation of a
 *    45-Qubit Quantum Circuit", SC17 (2017), arXiv:1704.01127.
 *  - Y. Suzuki et al., "Qulacs: a fast and versatile quantum circuit
 *    simulator for research purpose", Quantum 5, 559 (2021).
 */

import { QuantumState } from './quantum-state';
import { getModule } from './wasm-loader';

/** Diagnostic counts from {@link FusedCircuit.compile}.  Mirrors
 *  ``fuse_stats_t`` in ``src/optimization/fusion/fusion.h``. */
export interface FuseStats {
  /** Symbol-list length of the input circuit. */
  originalGates: number;
  /** Symbol-list length of the fused output. */
  fusedGates: number;
  /** Number of pair-wise 2x2 multiplications the fuser performed;
   *  equals (run-length - 1) summed across every fused run.  Zero
   *  when no single-qubit gate ran into a same-qubit successor
   *  before a multi-qubit barrier. */
  mergesApplied: number;
}

/** Tuple returned by {@link FusedCircuit.compile}. */
export interface FuseCompileResult {
  /** Owned fused-circuit handle; the caller must `dispose()` it. */
  fused: FusedCircuit;
  /** Summary statistics from the fuser pass. */
  stats: FuseStats;
}

type FusionModule = {
  _fuse_circuit_create: (n: number) => number;
  _fuse_circuit_free: (h: number) => void;
  _fuse_circuit_len: (h: number) => number;
  _fuse_circuit_num_qubits: (h: number) => number;
  _fuse_compile: (h: number, statsPtr: number) => number;
  _fuse_execute: (h: number, statePtr: number) => number;
  _malloc: (n: number) => number;
  _free: (p: number) => void;
  HEAPU32: Uint32Array;
} & Record<string, (...args: number[]) => number>;

/**
 * Owned WASM handle to a fusion-DAG circuit.
 *
 * Build the circuit with the fluent gate-append methods, then call
 * {@link compile} to produce a fused circuit and {@link execute} to
 * apply it to a {@link QuantumState}.  Call {@link dispose} to
 * release the underlying C handle.
 */
export class FusedCircuit {
  private handle: number;
  private mod: FusionModule;
  private nq: number;
  private freed = false;

  private constructor(handle: number, mod: FusionModule, numQubits: number) {
    this.handle = handle;
    this.mod = mod;
    this.nq = numQubits;
  }

  /** Allocate an empty fusion circuit on ``numQubits`` qubits. */
  static async create(numQubits: number): Promise<FusedCircuit> {
    if (numQubits < 1) {
      throw new RangeError(`numQubits must be >= 1, got ${numQubits}`);
    }
    const mod = (await getModule()) as unknown as FusionModule;
    const h = mod._fuse_circuit_create(numQubits);
    if (!h) throw new Error('fuse_circuit_create returned NULL');
    return new FusedCircuit(h, mod, numQubits);
  }

  /** Release the underlying WASM allocation. */
  dispose(): void {
    if (this.freed) return;
    this.mod._fuse_circuit_free(this.handle);
    this.freed = true;
  }

  /** Number of qubits the circuit was created for. */
  get numQubits(): number { return this.nq; }

  /** Number of symbolic gates currently in the circuit. */
  get length(): number {
    return this.mod._fuse_circuit_len(this.handle);
  }

  // ---- Single-qubit gates (fluent) --------------------------------------
  h(q: number): this   { this._call1('fuse_append_h', q);   return this; }
  x(q: number): this   { this._call1('fuse_append_x', q);   return this; }
  y(q: number): this   { this._call1('fuse_append_y', q);   return this; }
  z(q: number): this   { this._call1('fuse_append_z', q);   return this; }
  s(q: number): this   { this._call1('fuse_append_s', q);   return this; }
  sdg(q: number): this { this._call1('fuse_append_sdg', q); return this; }
  t(q: number): this   { this._call1('fuse_append_t', q);   return this; }
  tdg(q: number): this { this._call1('fuse_append_tdg', q); return this; }

  phase(q: number, theta: number): this {
    this._call1p('fuse_append_phase', q, theta); return this;
  }
  rx(q: number, theta: number): this {
    this._call1p('fuse_append_rx', q, theta); return this;
  }
  ry(q: number, theta: number): this {
    this._call1p('fuse_append_ry', q, theta); return this;
  }
  rz(q: number, theta: number): this {
    this._call1p('fuse_append_rz', q, theta); return this;
  }

  u3(q: number, theta: number, phi: number, lambda: number): this {
    this._guard1(q);
    const rc = this.mod._fuse_append_u3(this.handle, q, theta, phi, lambda);
    if (rc !== 0) throw new Error(`fuse_append_u3 rc=${rc}`);
    return this;
  }

  // ---- Two-qubit gates (fluent) -----------------------------------------
  cnot(ctrl: number, tgt: number): this {
    this._call2('fuse_append_cnot', ctrl, tgt); return this;
  }
  cz(ctrl: number, tgt: number): this {
    this._call2('fuse_append_cz', ctrl, tgt); return this;
  }
  cy(ctrl: number, tgt: number): this {
    this._call2('fuse_append_cy', ctrl, tgt); return this;
  }
  swap(a: number, b: number): this {
    this._call2('fuse_append_swap', a, b); return this;
  }

  cphase(ctrl: number, tgt: number, theta: number): this {
    this._call2p('fuse_append_cphase', ctrl, tgt, theta); return this;
  }
  crx(ctrl: number, tgt: number, theta: number): this {
    this._call2p('fuse_append_crx', ctrl, tgt, theta); return this;
  }
  cry(ctrl: number, tgt: number, theta: number): this {
    this._call2p('fuse_append_cry', ctrl, tgt, theta); return this;
  }
  crz(ctrl: number, tgt: number, theta: number): this {
    this._call2p('fuse_append_crz', ctrl, tgt, theta); return this;
  }

  // ---- Compile + execute -------------------------------------------------

  /** Run the run-length single-qubit fuser.  Returns an owned new
   *  {@link FusedCircuit} (caller must `dispose()`) plus a
   *  {@link FuseStats} summary. */
  compile(): FuseCompileResult {
    // fuse_stats_t = three size_t fields = 12 bytes on WASM32.
    const STATS_BYTES = 12;
    const statsPtr = this.mod._malloc(STATS_BYTES);
    try {
      const h = this.mod._fuse_compile(this.handle, statsPtr);
      if (!h) throw new Error('fuse_compile returned NULL');
      const idx = statsPtr >> 2;
      const stats: FuseStats = {
        originalGates: this.mod.HEAPU32[idx],
        fusedGates:    this.mod.HEAPU32[idx + 1],
        mergesApplied: this.mod.HEAPU32[idx + 2],
      };
      return { fused: new FusedCircuit(h, this.mod, this.nq), stats };
    } finally {
      this.mod._free(statsPtr);
    }
  }

  /** Apply this circuit to ``state`` in place.  Works on both fused
   *  and unfused circuits. */
  execute(state: QuantumState): void {
    const statePtr = state._internal_state_pointer();
    const rc = this.mod._fuse_execute(this.handle, statePtr);
    if (rc !== 0) throw new Error(`fuse_execute rc=${rc}`);
  }

  // ---- Internal helpers --------------------------------------------------

  private _guard1(q: number): void {
    if (this.freed) throw new Error('FusedCircuit is disposed');
    if (q < 0 || q >= this.nq) {
      throw new RangeError(`qubit ${q} out of [0, ${this.nq})`);
    }
  }

  private _guard2(a: number, b: number): void {
    if (this.freed) throw new Error('FusedCircuit is disposed');
    if (a < 0 || a >= this.nq || b < 0 || b >= this.nq) {
      throw new RangeError(`qubits ${a},${b} out of [0, ${this.nq})`);
    }
  }

  private _call1(fnName: string, q: number): void {
    this._guard1(q);
    const rc = this.mod[`_${fnName}`](this.handle, q);
    if (rc !== 0) throw new Error(`${fnName}(${q}) rc=${rc}`);
  }

  private _call1p(fnName: string, q: number, theta: number): void {
    this._guard1(q);
    const rc = this.mod[`_${fnName}`](this.handle, q, theta);
    if (rc !== 0) throw new Error(`${fnName}(${q},${theta}) rc=${rc}`);
  }

  private _call2(fnName: string, a: number, b: number): void {
    this._guard2(a, b);
    const rc = this.mod[`_${fnName}`](this.handle, a, b);
    if (rc !== 0) throw new Error(`${fnName}(${a},${b}) rc=${rc}`);
  }

  private _call2p(fnName: string, a: number, b: number, theta: number): void {
    this._guard2(a, b);
    const rc = this.mod[`_${fnName}`](this.handle, a, b, theta);
    if (rc !== 0) throw new Error(`${fnName}(${a},${b},${theta}) rc=${rc}`);
  }
}

File: bindings/rust/moonlab/src/fusion.rs
//! Single-qubit gate-fusion DAG (since v0.2.0; safe Rust wrapper
//! since v0.4.6).
//!
//! Wraps `src/optimization/fusion/fusion.{c,h}`.  Builds a symbolic
//! circuit, runs the run-length single-qubit fuser, and executes
//! the fused schedule on a [`crate::QuantumState`].  Collapses
//! adjacent single-qubit gates on the same qubit into one 2x2
//! matrix, dropping repeated full-state passes that state-vector
//! simulation pays for each gate.  On a five-layer
//! hardware-efficient ansatz at `n = 16` the fused execution is
//! roughly 2.2x faster than the unfused dispatch.
//!
//! Mirrors the v0.4.4 Python `moonlab.fusion.FusedCircuit`.
//!
//! ## Quick start
//!
//! ```no_run
//! use moonlab::{QuantumState, fusion::FusedCircuit};
//!
//! let mut circuit = FusedCircuit::new(4).unwrap();
//! circuit
//!     .h(0).unwrap()
//!     .rz(0, 0.3).unwrap()
//!     .rx(0, 0.7).unwrap()
//!     .cnot(0, 1).unwrap()
//!     .rz(1, 0.4).unwrap();
//!
//! let (fused, stats) = circuit.compile().unwrap();
//! assert!(stats.merges_applied >= 1);
//!
//! let mut state = QuantumState::new(4).unwrap();
//! fused.execute(&mut state).unwrap();
//! ```

use crate::error::{QuantumError, Result};
use crate::state::QuantumState;
use moonlab_sys::{
    fuse_append_cnot, fuse_append_cphase, fuse_append_crx, fuse_append_cry,
    fuse_append_crz, fuse_append_cy, fuse_append_cz, fuse_append_h,
    fuse_append_phase, fuse_append_rx, fuse_append_ry, fuse_append_rz,
    fuse_append_s, fuse_append_sdg, fuse_append_swap, fuse_append_t,
    fuse_append_tdg, fuse_append_u3, fuse_append_x, fuse_append_y,
    fuse_append_z, fuse_circuit_create, fuse_circuit_free, fuse_circuit_len,
    fuse_circuit_num_qubits, fuse_circuit_t, fuse_compile, fuse_execute,
    fuse_stats_t,
};
use std::ptr;

/// Diagnostic counts from [`FusedCircuit::compile`].  Mirrors
/// `fuse_stats_t` in the C header.
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct FuseStats {
    /// Symbol-list length of the input circuit.
    pub original_gates: usize,
    /// Symbol-list length of the fused output.
    pub fused_gates: usize,
    /// Number of pair-wise 2x2 multiplications the fuser performed;
    /// equals `(run_length - 1)` summed across every fused run.
    /// Zero when no single-qubit gate ran into a same-qubit
    /// successor before a multi-qubit barrier.
    pub merges_applied: usize,
}

/// Owned symbolic gate-fusion circuit.
pub struct FusedCircuit {
    handle: *mut fuse_circuit_t,
    num_qubits: usize,
}

// SAFETY: FusedCircuit owns its handle and is not shared across
// threads without external synchronization.
unsafe impl Send for FusedCircuit {}

impl FusedCircuit {
    /// Create an empty circuit on `num_qubits` qubits.
    pub fn new(num_qubits: usize) -> Result<Self> {
        if num_qubits == 0 {
            return Err(QuantumError::InvalidQubit { index: 0, max: 1 });
        }
        let handle = unsafe { fuse_circuit_create(num_qubits) };
        if handle.is_null() {
            return Err(QuantumError::AllocationFailed(num_qubits));
        }
        Ok(Self {
            handle,
            num_qubits,
        })
    }

    /// Number of qubits the circuit operates on.
    pub fn num_qubits(&self) -> usize {
        unsafe { fuse_circuit_num_qubits(self.handle) }
    }

    /// Number of symbolic gates currently in the circuit.
    pub fn len(&self) -> usize {
        unsafe { fuse_circuit_len(self.handle) }
    }

    /// True when no gates have been appended.
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    // ---- Single-qubit non-parameterised gates -----------------------------
    pub fn h(&mut self, q: i32)   -> Result<&mut Self> { self.call1("fuse_append_h",   unsafe { fuse_append_h  (self.handle, q) }) }
    pub fn x(&mut self, q: i32)   -> Result<&mut Self> { self.call1("fuse_append_x",   unsafe { fuse_append_x  (self.handle, q) }) }
    pub fn y(&mut self, q: i32)   -> Result<&mut Self> { self.call1("fuse_append_y",   unsafe { fuse_append_y  (self.handle, q) }) }
    pub fn z(&mut self, q: i32)   -> Result<&mut Self> { self.call1("fuse_append_z",   unsafe { fuse_append_z  (self.handle, q) }) }
    pub fn s(&mut self, q: i32)   -> Result<&mut Self> { self.call1("fuse_append_s",   unsafe { fuse_append_s  (self.handle, q) }) }
    pub fn sdg(&mut self, q: i32) -> Result<&mut Self> { self.call1("fuse_append_sdg", unsafe { fuse_append_sdg(self.handle, q) }) }
    pub fn t(&mut self, q: i32)   -> Result<&mut Self> { self.call1("fuse_append_t",   unsafe { fuse_append_t  (self.handle, q) }) }
    pub fn tdg(&mut self, q: i32) -> Result<&mut Self> { self.call1("fuse_append_tdg", unsafe { fuse_append_tdg(self.handle, q) }) }

    // ---- Single-qubit one-parameter gates ---------------------------------
    pub fn phase(&mut self, q: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_phase", unsafe { fuse_append_phase(self.handle, q, theta) })
    }
    pub fn rx(&mut self, q: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_rx", unsafe { fuse_append_rx(self.handle, q, theta) })
    }
    pub fn ry(&mut self, q: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_ry", unsafe { fuse_append_ry(self.handle, q, theta) })
    }
    pub fn rz(&mut self, q: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_rz", unsafe { fuse_append_rz(self.handle, q, theta) })
    }

    /// Universal single-qubit U3 gate.
    pub fn u3(&mut self, q: i32, theta: f64, phi: f64, lambda: f64)
        -> Result<&mut Self>
    {
        let rc = unsafe { fuse_append_u3(self.handle, q, theta, phi, lambda) };
        self.call1("fuse_append_u3", rc)
    }

    // ---- Two-qubit non-parameterised gates --------------------------------
    pub fn cnot(&mut self, ctrl: i32, tgt: i32) -> Result<&mut Self> {
        self.call1("fuse_append_cnot", unsafe { fuse_append_cnot(self.handle, ctrl, tgt) })
    }
    pub fn cz(&mut self, ctrl: i32, tgt: i32) -> Result<&mut Self> {
        self.call1("fuse_append_cz", unsafe { fuse_append_cz(self.handle, ctrl, tgt) })
    }
    pub fn cy(&mut self, ctrl: i32, tgt: i32) -> Result<&mut Self> {
        self.call1("fuse_append_cy", unsafe { fuse_append_cy(self.handle, ctrl, tgt) })
    }
    pub fn swap(&mut self, a: i32, b: i32) -> Result<&mut Self> {
        self.call1("fuse_append_swap", unsafe { fuse_append_swap(self.handle, a, b) })
    }

    // ---- Two-qubit one-parameter gates ------------------------------------
    pub fn cphase(&mut self, ctrl: i32, tgt: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_cphase", unsafe { fuse_append_cphase(self.handle, ctrl, tgt, theta) })
    }
    pub fn crx(&mut self, ctrl: i32, tgt: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_crx", unsafe { fuse_append_crx(self.handle, ctrl, tgt, theta) })
    }
    pub fn cry(&mut self, ctrl: i32, tgt: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_cry", unsafe { fuse_append_cry(self.handle, ctrl, tgt, theta) })
    }
    pub fn crz(&mut self, ctrl: i32, tgt: i32, theta: f64) -> Result<&mut Self> {
        self.call1("fuse_append_crz", unsafe { fuse_append_crz(self.handle, ctrl, tgt, theta) })
    }

    /// Run the single-qubit fuser.  Returns the new (fused) circuit
    /// together with a [`FuseStats`] summary.
    pub fn compile(&self) -> Result<(FusedCircuit, FuseStats)> {
        let mut stats = fuse_stats_t {
            original_gates: 0,
            fused_gates: 0,
            merges_applied: 0,
        };
        let h = unsafe { fuse_compile(self.handle, &mut stats) };
        if h.is_null() {
            return Err(QuantumError::AllocationFailed(0));
        }
        let fused = FusedCircuit {
            handle: h,
            num_qubits: self.num_qubits,
        };
        Ok((
            fused,
            FuseStats {
                original_gates: stats.original_gates,
                fused_gates: stats.fused_gates,
                merges_applied: stats.merges_applied,
            },
        ))
    }

    /// Apply the (fused or unfused) circuit to `state` in place.
    pub fn execute(&self, state: &mut QuantumState) -> Result<()> {
        let rc = unsafe { fuse_execute(self.handle, state.as_ptr()) };
        if rc != 0 {
            Err(QuantumError::Ffi(format!("fuse_execute rc={rc}")))
        } else {
            Ok(())
        }
    }

    // ---- Internal helpers --------------------------------------------------

    fn call1(&mut self, name: &'static str, rc: i32) -> Result<&mut Self> {
        if rc != 0 {
            Err(QuantumError::Ffi(format!("{name} rc={rc}")))
        } else {
            Ok(self)
        }
    }
}

impl Drop for FusedCircuit {
    fn drop(&mut self) {
        if !self.handle.is_null() {
            unsafe { fuse_circuit_free(self.handle) };
            self.handle = ptr::null_mut();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn create_and_len() {
        let c = FusedCircuit::new(4).unwrap();
        assert_eq!(c.num_qubits(), 4);
        assert_eq!(c.len(), 0);
        assert!(c.is_empty());
    }

    #[test]
    fn reject_zero_qubits() {
        assert!(FusedCircuit::new(0).is_err());
    }

    #[test]
    fn fluent_append_increments_len() {
        let mut c = FusedCircuit::new(3).unwrap();
        c.h(0).unwrap()
            .rz(0, 0.3).unwrap()
            .rx(0, 0.4).unwrap()
            .cnot(0, 1).unwrap()
            .ry(1, 0.5).unwrap();
        assert_eq!(c.len(), 5);
    }

    #[test]
    fn compile_run_fuses_three_into_one() {
        // h, rz, rx all on qubit 0 -> 1 FUSED_1Q + 2 merges.
        let mut c = FusedCircuit::new(2).unwrap();
        c.h(0).unwrap().rz(0, 0.3).unwrap().rx(0, 0.4).unwrap().cnot(0, 1).unwrap();
        let (fused, stats) = c.compile().unwrap();
        assert_eq!(stats.original_gates, 4);
        assert_eq!(stats.fused_gates, 2);
        assert_eq!(stats.merges_applied, 2);
        assert_eq!(fused.len(), 2);
    }

    #[test]
    fn compile_passthrough_when_no_runs() {
        let mut c = FusedCircuit::new(3).unwrap();
        c.h(0).unwrap().cnot(0, 1).unwrap().h(1).unwrap().cnot(1, 2).unwrap().h(2).unwrap();
        let (_fused, stats) = c.compile().unwrap();
        assert_eq!(stats.original_gates, 5);
        assert_eq!(stats.merges_applied, 0);
        assert_eq!(stats.fused_gates, 5);
    }

    #[test]
    fn execute_bell_state() {
        let mut c = FusedCircuit::new(2).unwrap();
        c.h(0).unwrap().cnot(0, 1).unwrap();
        let mut state = QuantumState::new(2).unwrap();
        c.execute(&mut state).unwrap();
        let probs = state.probabilities();
        assert!((probs[0] - 0.5).abs() < 1e-10);
        assert!(probs[1].abs() < 1e-10);
        assert!(probs[2].abs() < 1e-10);
        assert!((probs[3] - 0.5).abs() < 1e-10);
    }

    #[test]
    fn execute_fused_matches_unfused() {
        let build = || -> FusedCircuit {
            let mut c = FusedCircuit::new(3).unwrap();
            c.h(0).unwrap()
                .rz(0, 0.7).unwrap()
                .rx(0, 0.3).unwrap()
                .cnot(0, 1).unwrap()
                .ry(1, 0.2).unwrap()
                .rz(1, 0.9).unwrap()
                .cnot(1, 2).unwrap()
                .rx(2, 0.4).unwrap();
            c
        };

        let mut s_unfused = QuantumState::new(3).unwrap();
        build().execute(&mut s_unfused).unwrap();

        let mut s_fused = QuantumState::new(3).unwrap();
        let (fused, _) = build().compile().unwrap();
        fused.execute(&mut s_fused).unwrap();

        let pu = s_unfused.probabilities();
        let pf = s_fused.probabilities();
        for (u, f) in pu.iter().zip(pf.iter()) {
            assert!((u - f).abs() < 1e-10, "{u} != {f}");
        }
    }
}
